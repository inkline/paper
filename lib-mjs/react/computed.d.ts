import { Ref } from '@inkline/paper/types';
export declare function computed<T>(computeFn: () => T): Ref<T>;
