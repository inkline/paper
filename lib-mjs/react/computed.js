export function computed(computeFn) {
    return {
        get value() {
            return computeFn();
        }
    };
}
//# sourceMappingURL=computed.js.map