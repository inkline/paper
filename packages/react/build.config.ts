export default {
    entries: [
        {
            builder: 'mkdist',
            format: 'esm',
            ext: 'mjs',
            input: './src',
            declaration: true
        },
        {
            builder: 'mkdist',
            format: 'cjs',
            ext: 'cjs',
            input: './src',
            declaration: true
        }
    ]
};
