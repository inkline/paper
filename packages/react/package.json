{
  "name": "@inkline/paper-react",
  "description": "Inkline Paper React.js package.",
  "scripts": {
    "pre": "cp ../common/src/types.ts src/types.common.ts",
    "pretest": "npm run pre",
    "test": "vitest run",
    "prebuild": "npm run pre",
    "build": "tsc -d -p tsconfig.mjs.json && tsc -d -p tsconfig.cjs.json && mjscjs -p tsconfig.mjs.json && mjscjs -p tsconfig.cjs.json"
  },
  "devDependencies": {
    "@testing-library/react": "12.1.3",
    "@vitejs/plugin-react": "1.2.0",
    "jsdom": "19.0.0",
    "mjscjs": "1.0.9",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "rimraf": "3.0.2",
    "vite": "2.8.6",
    "vitest": "0.5.9"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/inkline/paper.git"
  },
  "license": "ISC"
}
