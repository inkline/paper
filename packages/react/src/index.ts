export * from './computed';
export * from './defineComponent';
export * from './definePlugin';
export * from './events';
export * from './fragment';
export * from './h';
export * from './helpers';
export * from './provide';
export * from './ref';
export * from './types';
export * from './watch';
