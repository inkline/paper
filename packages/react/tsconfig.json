{
    "extends": "../../tsconfig.json",
    "compilerOptions": {
        "types": [
            "@types/react"
        ]
    }
}
