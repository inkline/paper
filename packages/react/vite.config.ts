import { defineConfig } from 'vite';
import { resolve } from 'path';
import react from '@vitejs/plugin-react';

// https://vitejs.dev/config/
export default defineConfig({
    plugins: [
        react()
    ],
    resolve: {
        alias: [
            {
                find: /^@inkline\/paper\//,
                replacement: `${resolve(__dirname)}/src/`
            }
        ]
    },
    build: {
        lib: {
            entry: resolve(__dirname, 'src', 'index.ts'),
            name: 'InklinePaperReact',
            fileName: (format) => `react.${format}.js`
        },
        rollupOptions: {
            // make sure to externalize deps that shouldn't be bundled
            // into your library
            external: [
                'react'
            ],
            output: {
                exports: 'named',
                // Provide global variables to use in the UMD build
                // for externalized deps
                globals: {
                    react: 'React'
                }
            }
        }
    },
    test: {
        globals: true,
        environment: 'jsdom'
    }
});
