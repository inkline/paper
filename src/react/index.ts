export * from '@inkline/ucd/react/defineComponent';
export * from '@inkline/ucd/react/computed';
export * from '@inkline/ucd/react/h';
export * from '@inkline/ucd/react/helpers';
export * from '@inkline/ucd/react/ref';
export * from '@inkline/ucd/react/types';
