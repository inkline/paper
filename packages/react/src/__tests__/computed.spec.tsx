import { describe, it, expect } from 'vitest';
import { fireEvent, render } from '@testing-library/react';
import { computed, defineComponent, h, ref } from '../index';

describe('react', () => {
    describe('computed()', () => {
        it('should create a new computed value', () => {
            const Component = defineComponent({
                setup () {
                    const state = computed(() => 'hello');

                    expect(state).toEqual({
                        value: 'hello'
                    });

                    return { state };
                },
                render ({ state }) {
                    return h('div', {}, [state.value]);
                }
            });

            const wrapper = render(h(Component) as any);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        it('should compute value based on other refs', () => {
            const Component = defineComponent({
                setup () {
                    const a = ref(1);
                    const b = ref(2);
                    const state = computed(() => a.value + b.value);

                    expect(state.value).toEqual(3);

                    return { state };
                },
                render ({ state }) {
                    return h('div', {}, [state.value]);
                }
            });

            const wrapper = render(h(Component) as any);
            expect(wrapper.container.firstChild).toMatchSnapshot();
        });

        it('should update value based on other refs', async () => {
            const Component = defineComponent({
                setup () {
                    const a = ref(1);
                    const b = ref(2);
                    const state = computed(() => a.value + b.value);

                    const onClick = () => {
                        a.value = 4;
                        b.value = 6;
                    };

                    return { state, onClick };
                },
                render ({ state, onClick }) {
                    return h('div', { onClick }, [state.value]);
                }
            });

            const wrapper = render(h(Component) as any);
            await fireEvent.click(wrapper.container.firstChild as Element);

            expect(wrapper.container.firstChild).toMatchSnapshot();
        });
    });
});
