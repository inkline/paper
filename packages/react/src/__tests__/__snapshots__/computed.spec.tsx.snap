// Vitest Snapshot v1

exports[`react > computed() > should compute value based on other refs 1`] = `
<div>
  3
</div>
`;

exports[`react > computed() > should create a new computed value 1`] = `
<div>
  hello
</div>
`;

exports[`react > computed() > should update value based on other refs 1`] = `
<div>
  10
</div>
`;
